lightning-invoice/src/lib.rs

```diff
@@ -890,6 +892,76 @@ impl Deref for PositiveTimestamp {
 }
 
 impl Invoice {
+	/// Utility to construct an invoice. `amt_msat` can be 0 for 0-value invoices.
+	pub fn from_channelmanager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(
+		channelmanager: &ChannelManager<Signer, M, T, K, F, L>, amt_msat: u64, description: String,
+		network: Network,
+	) -> Result<Invoice, CreationError>
+	where
+		M::Target: 'static + chain::Watch<Signer>,
+		T::Target: 'static + BroadcasterInterface,
+		K::Target: 'static + KeysInterface<Signer = Signer>,
+		F::Target: 'static + FeeEstimator,
+		L::Target: 'static + Logger,
+	{
+		// Marshall route hints.
+		let our_channels = channelmanager.list_usable_channels();
+		let mut route_hints = vec![];
+		let mut min_final_cltv_expiry = 18;
+		for channel in our_channels {
+			let short_channel_id = match channel.short_channel_id {
+				Some(id) => id,
+				None => continue,
+			};
+			let forwarding_info = match channel.counterparty_forwarding_info {
+				Some(info) => info,
+				None => continue,
+			};
+			if forwarding_info.cltv_expiry_delta > min_final_cltv_expiry {
+				min_final_cltv_expiry = forwarding_info.cltv_expiry_delta;
+			}
+			route_hints.push(vec![RouteHintHop {
+				src_node_id: channel.remote_network_id,
+				short_channel_id,
+				fees: RoutingFees {
+					base_msat: forwarding_info.fee_base_msat,
+					proportional_millionths: forwarding_info.fee_proportional_millionths,
+				},
+				cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+				htlc_minimum_msat: None,
+				htlc_maximum_msat: None,
+			}]);
+		}
+
+		let (payment_hash, payment_secret) = channelmanager.get_payment_secret_preimage(
+			if amt_msat == 0 { None } else { Some(amt_msat) },
+			min_final_cltv_expiry.into(),
+			0,
+		);
+		let secp_ctx = Secp256k1::new();
+		let our_node_pubkey = channelmanager.get_our_node_id();
+		let mut invoice = InvoiceBuilder::new(match network {
```

Comment:
Needs features, right, or is just waiting on #876 to land?

Reply:
Yeah, locally it's based on features. Now getting the sample based on this PR so I can add a test with an invoice I actually know is good...